### PR TITLE
Create API-Testing-Guidelines.md

### LINKED ISSUE #94: Test definitions - implementations guidelines and artifacts
**Problem description**
Delivery of  **API test cases and documentation** is one of criteria in API Readiness Checklist.
Each subproject needs to develop a Gherkin feature file for specified API.
_If subprojects also intend to add test implementations, an aligned single implementation that is agreed amongst all provider implementors could be added to the main subproject repo._

**Possible evolution**
Definition of common guidelines and artifacts that would simplify development of test definition, documentation and implementation based on current experience in subprojects and within participating organizations.

**Additional context**
As listed in https://github.com/camaraproject/SimSwap/issues/63  following projects  already provide Gherkin feature files:

- [EdgeCloud (Traffic Influence)](https://github.com/camaraproject/EdgeCloud/blob/main/code/Test_definitions/Traffic_Influence_Test.feature)
 -  [QualityOnDemand](https://github.com/camaraproject/QualityOnDemand/blob/main/code/cucumber/src/test/resources/feature/QoD_API_Test.feature)
   - [HomeDevicesQoD](https://github.com/camaraproject/HomeDevicesQoD/blob/main/code/Test_definitions/home_devices_qod_setQos.feature)

### PR DESCRIPTION
**Pull request by shilpa-padgaonkar**
#### What type of PR is this?

Add one of the following kinds:
* documentation

#### What this PR does / why we need it:
Adds API-testing-guidelines doc

#### Which issue(s) this PR fixes:

Fixes #94 

#### Changelog input

```
 release-note

```

#### Additional documentation 

This section can be blank.

```
docs

```

### CODE REVIEW

**Comment by shilpa-padgaonkar** [comment]
@bigludo7 @jlurien @mdomale: Kindly review if all your points have been covered in the doc.

**Comment by mdomale** [comment]
@shilpa-padgaonkar Document looks fine to me.

**Review by patrice-conil — COMMENTED** [review]
/LGTM
Maybe a link to subproject  sample could help? This one for exemple
[Submission of test cases by trehman-gsma · Pull Request #68 · camaraproject/SimSwap (github.com](https://github.com/camaraproject/SimSwap/pull/68/files)

**Comment by shilpa-padgaonkar** [comment]
@patrice-conil : made the fix as you suggested. I simply used qod as example because simswap example is still a PR

**Comment by bigludo7** [comment]
Hello @shilpa-padgaonkar 
First, thanks for the effort :)

Like my colleague @patrice-conil  I think referring a sample is very useful.

I have an issue with QoD file because it not reflect the rules (several when/then, name). I've applied rules here: https://github.com/camaraproject/DeviceLocation/pull/119#pullrequestreview-1818257946 - Could you please take a look on this proposal and check if it aligned with your view? 

We should have an exemple fully aligned with these rules i guess and use it as base.

**Comment by shilpa-padgaonkar** [comment]
@thanks Ludovic for your quick feedback.  I simply added qod as an example for the location (where should the feature file reside). 
But you are right, a link to a sample which is based on these guidelines would be a very good add. I have looked at your feature file and it LGTM. I will also request @mdomale to briefly take a look and provide quick feedback on the file.

BTW, should we add your example link already, or would you want to do this after your PR gets merged (I am fine both ways)?

**Comment by mdomale** [comment]
@bigludo7  Except minor comments , mentioned feature file in PR https://github.com/camaraproject/DeviceLocation/pull/119#pullrequestreview-1818257946
 LGTM   @shilpa-padgaonkar

**Comment by bigludo7** [comment]
Thanks a lot @shilpa-padgaonkar and @mdomale  for your review.

Regarding:

> BTW, should we add your example link already, or would you want to do this after your PR gets merged (I am fine both ways)?

I suggest we wait from Device location project feedback but I will inform then that current proposal is "OK" from our perpsective. Adding @jlurien  in the loop.

File: documentation/API-Testing-Guidelines.md
Comment (jlurien) [inline documentation/API-Testing-Guidelines.md:14]:
```suggestion
Based on the decision taken in the Commonalities working group and as documented in [API-Readiness-Checklist](https://github.com/camaraproject/Commonalities/blob/main/documentation/API-Readiness-Checklist.md), at least one Gherkin feature file will be added to the main subproject repo and this will fulfill the minimum criteria of readiness with respect to API test cases and documentation. If subprojects also intend to add test implementations, an aligned single implementation that is agreed among all provider implementors could be added to the main subproject repo. If no alignment is possible, each provider implementer will add the test implementation to their own repos.
```

Other comments:

- In some cases we may need several Gherkin files, as in the issue we discussed the convenience to have a .feature file per API operation, specially for APIs with many endpoints.
- Not sure about the convenience to add implementations directly to the subprojects

File: documentation/API-Testing-Guidelines.md
Comment (jlurien) [inline documentation/API-Testing-Guidelines.md:25]: We should not impose one single .feature file per API, but one feature file per operation or API endpoint/resource. In general, it is not advisable to test several unrelated "API features" within the same .feature file. For closely related operations we may add scenarios in the same file testing crossed behaviour between several operations.

File: documentation/API-Testing-Guidelines.md
Comment (jlurien) [inline documentation/API-Testing-Guidelines.md:39]: Where does this reference recommend having one file per API? APIs are not explicitly mentioned, but considering several  files for specific parts is advised:

> How should the files be separated?
> Because separating the features might result in a large number of files, you must consider how to separate the features into various files.
> 
> A popular solution is to have a file with features that collect everything relevant to a specific part of the program and even organize it into directories.

In the issue #94, another [reference](https://copyprogramming.com/howto/multiple-feature-inside-single-feature-file#multiple-feature-inside-single-feature-file) was mentioned. There, it is advised to:

> Putting multiple features in a single feature file is not recommended. It is recommended to have a separate feature file for each feature. However, it is acceptable to include multiple scenarios in a single feature file.

So we should identify what are the "features" in the context of an API, and put a feature per file. To me a feature in this context is an API operation or resource. For simple APIs with single operations, it is quite straight-forward to decide that one file is enough, e.g. location-verification API, but for more complex APIs, such as QoD, we may need several files to test unrelated features, e.g. discovery of QoD profiles, creating QoD sessions, etc

Comment (mdomale) [reply documentation/API-Testing-Guidelines.md:39]: @jlurien  The reference link for one feature per API is https://copyprogramming.com/howto/multiple-feature-inside-single-feature-file#multiple-feature-inside-single-feature-file @shilpa-padgaonkar

**Comment by shilpa-padgaonkar** [comment]
@jlurien : Thanks for your comments. I will fix the incorrect link with the new one provided by @mdomale and add it to the doc.
As you have given the example of qod api to support the case of using multiple feature files (feature in your context is an API operation or resource), I would like to ask for some additional feedback (especially from the qod codeowners and/or maintainers) to get more view points on this topic.
Tagging here the codeowners: @hdamker @RandyLevensalor @eric-murray

**Comment by bigludo7** [comment]
@shilpa-padgaonkar I tend to agree with @jlurien that we should flex the following sentence:

> One feature file per API is advisable so that all scenarios can be covered corresponding to the API & the corresponding resource.

Indeed, as I'm involved in Device Status for example, we have there one API (one yaml) that feature 3 distincts capabilities:

- Get on the fly if the mobile is in roaming situation
- Get on the fly if the mobile is reacheable (connectivity)
- Subscription to get notification when mobile change roaming/connectivity state

_(the granularity of this API is perhaps challengeable but that's another topic)_

I assume that it will be in this case simpler to have 3 dedicated features --> I will trigger the discussion in this project.

Let's see the feedback we have from QoD & Device Status... Perhaps we can put in like this: Granularity of the feature file should be at the project level but it is recommended to group in one file all scenario for one given API fonction. If a CAMARA API features several independent fonctions it is recommended to provide several files.

**Comment by shilpa-padgaonkar** [comment]
@bigludo7 : Thanks Ludovic for driving this in devicestatus. :)
Now awaiting feedback from devicestatus and qod

File: documentation/API-Testing-Guidelines.md
Comment (jlurien) [reply documentation/API-Testing-Guidelines.md:39]: @mdomale. In my understanding, the guidelines there does not advice for one feature per API but one "feature" per *.feature file. So that is what I suggest, to identify which are the "features" to test in the context of an API, and put each feature in each own *.feature file. Thus, an API may have several *.feature files for testing different aspects of the API. In simple APIs with just one endpoint that will not be the case.

**Comment by jlurien** [comment]
I agree with @bigludo7 and Device Status is a very good example. "Feature" is a very abstract concept, but the common advice is to not mix unrelated functionality in the same file. It is more easy to work with several small files dedicated to test an specific part of an application or API, in our case, that having only a  huge file with hundreds of unrelated scenarios. 
Cucumber and other frameworks accept a folder as input and are prepared to look for and execute several *.feature files in one run. So I'm in favor of relaxing the guidelines and let each subproject decide on how many "features" can be identified for some specific API.

**Comment by shilpa-padgaonkar** [comment]
@jlurien and @bigludo7 : Could you also kindly propose a text in the meanwhile which would reflect this more relaxed guideline, as opposed to "One feature file per API is advisable so that all scenarios can be covered corresponding to the API".

As we need to agree and merge this document for the upcoming release, I would request you to already prep for this text and once we get this similar feedback as suggested above from devicestatus and qod, we can add your text and proceed with approving and merging this doc.

File: documentation/API-Testing-Guidelines.md
Comment (bigludo7) [reply documentation/API-Testing-Guidelines.md:25]: @shilpa-padgaonkar following today Commonalities meeting I proposed to update this sentence but I'm not able to update the file. My proposal for line 25: 
Granularity of the feature file must be decided at the project level but it is recommended to:

- group in one file all scenarios for one given API capability (that can cover several endpoints). 
- provide several files when one CAMARA API (yaml) covers several independent functions that can be provided independently

Perhaps @jlurien  could provide a better wording. I think we're aligned.

**Comment by bigludo7** [comment]
> @jlurien and @bigludo7 : Could you also kindly propose a text in the meanwhile which would reflect this more relaxed guideline, as opposed to "One feature file per API is advisable so that all scenarios can be covered corresponding to the API".
> 
> As we need to agree and merge this document for the upcoming release, I would request you to already prep for this text and once we get this similar feedback as suggested above from devicestatus and qod, we can add your text and proceed with approving and merging this doc.

I've just provided a sentence as a review comment ;)

**Comment by shilpa-padgaonkar** [comment]
@bigludo7 : Thanks. Your suggestions are now added to the doc :)

**Review by bigludo7 — APPROVED** [review]
LGTM

**Review by RandyLevensalor — APPROVED** [review]

**Comment by shilpa-padgaonkar** [comment]
Requesting @jlurien to kindly review and approve

File: documentation/API-Testing-Guidelines.md
Comment (jlurien) [inline documentation/API-Testing-Guidelines.md:26]:
```suggestion
    -	group in one file all scenarios testing one closely related API capability (that can cover one or several endpoints).
```

**Review by jlurien — APPROVED** [review]
LGTM with some minor suggestion

**Comment by shilpa-padgaonkar** [comment]
Thanks all for the contributions, review and approval. @rartych: Could you kindly check that there are no more open points ignored as a part of the review and if not, proceed to merge the PR?

**Comment by rartych** [comment]
@RubenBG7 @jordonezlucena - as @jlurien approved the PR, could you approve it as codewoners?

**Review by rartych — APPROVED** [review]

**Pull request merged**
